Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/WebHDFSDeploy.java

```diff
@@ -0,0 +1,193 @@
+import com.microsoft.azure.hdinsight.sdk.storage.webhdfs.WebHdfsParamsBuilder;
+import com.microsoft.azure.hdinsight.spark.jobs.JobUtils;
+import com.microsoft.azure.sqlbigdata.sdk.cluster.SqlBigDataLivyLinkClusterDetail;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import com.microsoft.azuretools.azurecommons.helpers.Nullable;
+import org.apache.commons.httpclient.methods.PutMethod;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.entity.BufferedHttpEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.InputStreamEntity;
+import org.apache.http.message.BasicNameValuePair;
+import rx.Observable;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.UnknownServiceException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class WebHDFSDeploy implements Deployable, ILogger {
+    @NotNull
+    private IClusterDetail cluster;
+
+    @NotNull
+    private String webHDFSRootPath;
+
+    @Nullable
+    private HttpObservable http;
+
+    @NotNull
+    private List<NameValuePair> createDirReqParams;
+
+    @NotNull
+    private List<NameValuePair> uploadReqParams;
+
+    public WebHDFSDeploy(IClusterDetail cluster, String webHDFSRootPath) {
+        this.cluster = cluster;
+        this.webHDFSRootPath = webHDFSRootPath;
+        this.uploadReqParams = new WebHdfsParamsBuilder("CREATE")
+                .setOverwrite("true")
+                .build();
+
+        this.createDirReqParams = new WebHdfsParamsBuilder("MKDIRS")
+                .setPermission("755")
+                .build();
+
+        try {
+            http = createHttpObservable(cluster);
+        } catch (HDIException e) {
+            log().warn(e.getMessage());
+            http = null;
+        }
+
+        if (http != null) {
+            http.setDefaultRequestConfig(RequestConfig.custom().setExpectContinueEnabled(true).build());
+        } else {
+            throw new IllegalArgumentException("Cluster type doesen't support webHDFS storage type.");
+        }
+    }
+
+    @NotNull
+    public Observable<String> deploy(@NotNull String artifactPath) {
+        File file = new File(artifactPath);
+        URI remote = URI.create(webHDFSRootPath)
+                .resolve(JobUtils.getFormatPathByDate() + "/");
+
+        //three steps to upload via webhdfs
+        // 1.put request to create new dir
+        // 2.put request to get 307 redirect uri from response
+        // 3.put redirect request with file content as setEntity
+        try {
+            HttpPut req = new HttpPut(remote.toString());
+            http.request(req, null, this.createDirReqParams, null)
```

Comment:
Why not use `http.put()`?

Reply:
1.put method calls toStringOnlyOkResponse which treats statuscode >200 as invalid but we need use the 307 response 2.  toStringOnlyOkResponse  calls getResultFromHttpResponse which miss the header info in response but we need use the  'Location' in the response

---
In reply to: [252106151](https://github.com/Microsoft/azure-tools-for-java/pull/2619#discussion_r252106151) [](ancestors = 252106151)